examples/react-dapp/src/App.tsx

```diff
@@ -18,52 +21,138 @@ type Role = {
+      } catch (err) {
+        console.error(err)
+      }
+    }
+    setIsLoading(false);
+  }
+
+  const loginWithWalletConnect = async function () {
+    setIsLoading(true);
+    const { did } = await iam.initializeConnection();
+    if (did) {
+      setDID(did)
+      try {
+        await verifyIdentity();
+      } catch (err) {
+        console.error(err)
+      }
     }
-    const { data: roles } = await axios.get<Role[]>('/roles', config)
-    const { data: user } = await axios.get<User>('/user', config)
+    setIsLoading(false);
+  }
 
-    setRoles(roles)
+  const logout = function () {
+    setDID('');
   }
-  return (
-    <div className="App">
-      <header className="App-header">
+
+  const renderLoading = () => {
+    return (
+      <div>
+        <Spinner />
+        <span>
+          Loading... (Please sign messages using your connected wallet)
+        </span>
+      </div>
+    )
+  }
+
+  const renderLoginResults = () => {
+    return (
+      <div>
+        <p>Hello user!</p>
+        <p>
+          Your decentralised identifier: <br />
+          <strong>{did}</strong>
+        </p>
         {roles && roles.length > 0 ? (
-          <>
-            <Typography>User roles:</Typography>
+          <div className="rolesContainer">
+            <p>These are your validated roles:</p>
             {roles.map(({ name, namespace }) => (
-              <p key={namespace}>{`${name} at ${namespace}`}</p>
+              <p key={namespace}><strong>{`${name}`}</strong>{` at ${namespace}`}</p>
             ))}
-          </>
-        ) : null}
-        <Button variant="contained" color="primary" onClick={handleLogin}>
-          login
-        </Button>
-      </header>
-    </div>
+          </div>
+        )
+          : (
+            <div>
+              You do not have any issued role at the moment, please login into switchboard and search for
+              apps, orgs to enrol.
+            </div>
+          )}
+        <div className="logoutContainer">
+          <button onClick={logout} className="button">
+            <span>Logout</span>
+          </button>
+        </div >
+      </div >
+    )
+  }
+
+  const renderLoginOptions = () => {
+    return (
+      <div className="container">
+        <button className="button" onClick={async () => await loginWithWalletConnect()}>
+          <img alt="walletconnect logo" className="walletconnect" src={walletconnectIcon} />
+          <span>Login with Wallet Connect</span>
+        </button>
+        <button className="button" onClick={async () => await login({ useMetamaskExtension: true })}>
+          <img alt="metamask logo" className="metamask" src={metamaskLogo} />
+          <span>Login with Metamask</span>
+        </button>
+      </div>
+    )
+  }
+
+  const loginJsx = () => {
+    if (isLoading) {
+      return renderLoading()
+    }
+    else if (did) {
+      return renderLoginResults();
+    }
+    else {
+      return renderLoginOptions();
+    }
+  };
```

Comment:
@dwojno Is this what you had in mind for the `loginJsx` refactor or is there a further improvement?

Reply:
Yes, perfect 👍